Source/Csla/Configuration/Fluent/DataPortalConfigurationExtensions.cs

```diff
@@ -58,7 +58,11 @@ public static DataPortalClientOptions AddServerSideDataPortal(this DataPortalCli
       options?.Invoke(opt);
       var services = builder.Services;
       services.TryAddScoped(typeof(IAuthorizeDataPortal), opt.AuthorizerProviderType);
-      services.TryAddScoped((p) => opt.InterceptorProviders);
+      foreach (Type interceptorType in opt.InterceptorProviders)
```

Comment:
Won't this throw an exception when more than one interceptor is provided? I don't think you can register more than one entry for a type. This is why I registered the list of providers, so the interceptormanager could unpack the list.

Reply:
No, it doesn't. TryAdd is slightly a misnomer, and there is a special behaviour for multiple registrations.

If you use Add twice, the first registration sort of overwrites the first, and there is no exception. TryAdd checks if an existing registration exists and avoids registering if there is one, so as not to 'overwrite' the first. In my opinion it's a slightly confusing name, because it's not used in the sense that Try is used in the rest of the framework.

However, in reality, I don't think Add _actually_ overwrites a registration anyway. I think both registrations are stored, but if you ask for a single instance then you get the most recent. However, if you ask for an IEnumerable, then you get an instance of each one registered.

The code I have shown does work, but it works because InterceptorManager receives an IEnumerable<IInterceptDataPortal> on its constructor. It's a really nice feature of DI, but it seems to be one that is not all that well known. I only know about it because I chanced upon it whilst researching this very feature. I think I was lucky; if I'd got a different answer to a very slightly different Google search, I'd not have found it.

Reply:
I tried running the code you used, and I got back an empty enumerator. I think that is mainly because it isn't associated with the interface I am asking for.

Reply:
I suppose for completeness I should call out that the IEnumerable<T> approach in constructors works for zero or more registrations. If nothing is registered it returns an empty enumerator. It's a really nice, straightforward approach because of that behaviour - a simple way to add optional extensibility using a pipeline-style (or decorator-style) pattern.